File: sdks/flutter/lib/src/astro_client.dart
import 'astro_http.dart';
import 'payments/payments_client.dart';
import 'alias/alias_client.dart';
import 'openbanking/openbanking_client.dart';
import 'identity/identity_client.dart';
import 'webhooks/webhook_verifier.dart';

export 'models/models.dart';
export 'payments/payments_client.dart';
export 'alias/alias_client.dart';
export 'openbanking/openbanking_client.dart';
export 'identity/identity_client.dart';
export 'webhooks/webhook_verifier.dart';

class AstroConfig {
  final String baseUrl;
  final String? merchantKey;
  final String? bankKey;
  final String? adminKey;
  final Duration timeout;

  const AstroConfig({
    required this.baseUrl,
    this.merchantKey,
    this.bankKey,
    this.adminKey,
    this.timeout = const Duration(seconds: 30),
  });
}

class AstroClient {
  late final PaymentsClient payments;
  late final AliasClient alias;
  late final OpenBankingClient openBanking;
  late final IdentityClient identity;

  AstroClient(AstroConfig config) {
    final http = AstroHttpClient(
      baseUrl: config.baseUrl,
      merchantKey: config.merchantKey,
      bankKey: config.bankKey,
      adminKey: config.adminKey,
      timeout: config.timeout,
    );
    payments = PaymentsClient(http);
    alias = AliasClient(http);
    openBanking = OpenBankingClient(http);
    identity = IdentityClient(http);
  }

  WebhookVerifier webhookVerifier(String secret) => WebhookVerifier(secret);
}

File: sdks/swift/Sources/AstroSDK/AstroClient.swift
import Foundation

public final class AstroClient {

    public let payments: PaymentsClient
    public let alias: AliasClient
    public let openBanking: OpenBankingClient
    public let identity: IdentityClient

    private let http: AstroHTTPClient

    public init(config: AstroConfig) {
        http = AstroHTTPClient(config: config)
        payments = PaymentsClient(http: http)
        alias = AliasClient(http: http)
        openBanking = OpenBankingClient(http: http)
        identity = IdentityClient(http: http)
    }

    public func webhookVerifier(secret: String) -> WebhookVerifier {
        WebhookVerifier(secret: secret)
    }
}

// MARK: - HTTP Engine

final class AstroHTTPClient {
    let config: AstroConfig
    private let session: URLSession
    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()

    init(config: AstroConfig) {
        self.config = config
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = config.timeout
        self.session = URLSession(configuration: cfg)
    }

    func url(_ path: String, query: [String: String?]? = nil) -> URL {
        var components = URLComponents(url: config.baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if let q = query {
            components.queryItems = q.compactMap { k, v in v.map { URLQueryItem(name: k, value: $0) } }
        }
        return components.url!
    }

    func request(_ url: URL, method: String = "GET", body: Encodable? = nil, extraHeaders: [String: String] = [:]) -> URLRequest {
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let key = config.merchantKey { req.setValue("Bearer \(key)", forHTTPHeaderField: "Authorization") }
        if let key = config.bankKey { req.setValue(key, forHTTPHeaderField: "X-OpenWave-Bank-Key") }
        if let key = config.adminKey { req.setValue(key, forHTTPHeaderField: "X-OpenWave-Admin-Key") }
        extraHeaders.forEach { req.setValue($0.value, forHTTPHeaderField: $0.key) }
        if let body = body { req.httpBody = try? JSONEncoder().encode(body) }
        return req
    }

    func perform<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw AstroError(statusCode: 0, code: "NETWORK_ERROR", message: "No HTTP response")
        }
        if (200..<300).contains(http.statusCode) {
            return try decoder.decode(T.self, from: data)
        }
        let errorBody = try? JSONDecoder().decode([String: [String: String]].self, from: data)
        let err = errorBody?["error"]
        throw AstroError(
            statusCode: http.statusCode,
            code: err?["code"] ?? "UNKNOWN_ERROR",
            message: err?["message"] ?? "HTTP \(http.statusCode)",
            detail: err?["detail"]
        )
    }
}
